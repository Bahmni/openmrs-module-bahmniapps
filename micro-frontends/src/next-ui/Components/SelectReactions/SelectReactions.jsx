import React, { useEffect, useState } from "react";
import propTypes from "prop-types";
import "../../../styles/common.scss";
import { Checkbox, Search, Tag } from "carbon-components-react";
import { FormattedMessage } from "react-intl";
import { cloneDeep } from "lodash";

const getReactionIds = (reactions) => {
  return Object.keys(reactions);
};

export const SelectReactions = (props) => {
  const { onChange, reactions, selectedAllergen } = props;
  const initialReactionIds = getReactionIds(reactions);
  const [searchResults, setSearchResults] = useState(initialReactionIds);
  const [isSearchResultEmpty, setIsSearchResultEmpty] = useState(true);
  const [selectedReactions, setSelectedReactions] = useState([]);
  const [allReactions] = useState(cloneDeep(reactions));
  const [searchKey, setSearchKey] = useState("");

  const search = (key) => {
    setSearchKey(key)
    if (!key) {
      setIsSearchResultEmpty(true);
      setSearchResults(initialReactionIds);
      return;
    }
    const searchIds = [];
    const secondarySearchIds = [];
    Object.keys(allReactions).forEach((reactionId) => {
      if (
        allReactions[reactionId].name
          .toLowerCase()
          .startsWith(key.toLowerCase())
      ) {
        searchIds.push(reactionId);
      } else if (
        allReactions[reactionId].name.toLowerCase().includes(key.toLowerCase())
      ) {
        secondarySearchIds.push(reactionId);
      }
    });
    searchIds.push(...secondarySearchIds);
    if (searchIds.length === 0) {
      setIsSearchResultEmpty(true);
      setSearchResults(initialReactionIds);
    } else {
      setIsSearchResultEmpty(false);
      setSearchResults(searchIds);
    }
  };
  useEffect(() => {
    if(selectedReactions.length > 0){
      setSearchKey("")
    }
    onChange(selectedReactions);
  }, [selectedReactions]);

  return (
    <div className={"section-next-ui"}>
      <div className={"font-large selected-allergen"}>
        <FormattedMessage id={"SELECTED_ALLERGEN"} defaultMessage={"Selected Allergen:"}/> {selectedAllergen.name}</div>
      <div className={"font-large bold"}><FormattedMessage id={"SEARCH_REACTION"} defaultMessage={"Search Reaction"}/><span className={"red-text"}>&nbsp;*</span></div>
      <div>
        <Search
          id={"reaction-search"}
          placeholder={"Type to search Reactions"}
          value={searchKey}
          onChange={(e) => {
            search(e.target.value);
          }}
        />
      </div>
      {selectedReactions.length > 0 && (
        <div>
          {selectedReactions.map((reactionId) => {
            return (
              <Tag
                key={reactionId}
                filter={true}
                type={"blue"}
                onClose={() => {
                  allReactions[reactionId].isSelected = false;
                  setSelectedReactions(
                    selectedReactions.filter(
                      (reaction) => reaction !== reactionId
                    )
                  );
                }}
              >
                {allReactions[reactionId].name}
              </Tag>
            );
          })}
        </div>
      )}
      <div>
        {isSearchResultEmpty && (
          <div className={"font-small selected-allergen"}>
            <FormattedMessage id={"COMMON_REACTIONS"} defaultMessage={"Common Reactions"}/>
          </div>
        )}
        {searchResults.map((reactionId) => {
          return (
            <Checkbox
              id={reactionId}
              key={reactionId}
              labelText={allReactions[reactionId].name}
              checked={allReactions[reactionId].isSelected}
              onChange={(e) => {
                allReactions[reactionId].isSelected = e;
                if (e) {
                  setSelectedReactions((oldReactions) => [
                    ...oldReactions,
                    reactionId,
                  ]);
                } else {
                  setSelectedReactions(
                    selectedReactions.filter(
                      (reaction) => reaction !== reactionId
                    )
                  );
                }
              }}
            />
          );
        })}
      </div>
    </div>
  );
};

SelectReactions.propTypes = {
  onChange: propTypes.func.isRequired,
  reactions: propTypes.object.isRequired,
  selectedAllergen: propTypes.object.isRequired,
};
