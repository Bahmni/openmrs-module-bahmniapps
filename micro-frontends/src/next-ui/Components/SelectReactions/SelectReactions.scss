.selected-allergen{
  margin-bottom: 8px;
  margin-top: 10px;
}
