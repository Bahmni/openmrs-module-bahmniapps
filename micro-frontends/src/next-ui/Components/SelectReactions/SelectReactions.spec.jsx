import React from "react";
import { fireEvent, render, screen, act } from "@testing-library/react";
import { SelectReactions } from "./SelectReactions";

describe("Select reactions", function () {
  const onChange = jest.fn();
  const mockReactions = {
    "101AA": { name: "GI Upset" },
    "102AA": { name: "Fever" },
    "103AA": { name: "Headache" },
    "104AA": { name: "Nausea" },
    "105AA": { name: "Cough" },
  };

  const mockSelectedAllergen = {
    name: "Eggs",
    kind: "Food",
    uuid: "162301AAAAAA",
  };

  const selectReaction = (container) => {
    const searchInput = container.querySelector(".bx--search-input");
    fireEvent.change(searchInput, { target: { value: "GI" } });
    const checkbox = screen.getByLabelText("GI Upset");
    fireEvent.click(checkbox);
  };
  it("should render SelectReactions", function () {
    const { container } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("should remove chiclets when reaction checkbox is unselected", function () {
    const { container, getAllByText } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );

    selectReaction(container);
    const tag = container.querySelector(".bx--tag");
    const checkbox = screen.getAllByLabelText("GI Upset")[1];
    expect(tag).toBeTruthy();
    expect(checkbox.checked).toEqual(true);
    expect(getAllByText("GI Upset").length).toEqual(2);

    //unselect checkbox
    fireEvent.click(checkbox);
    expect(container.querySelector(".bx--tag")).toBeNull();
    expect(getAllByText("GI Upset").length).toEqual(1);
  });

  it("should unselected checkbox when corresponding Tag is closed", function () {
    const { container, getAllByText } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    selectReaction(container);
    const tag = container.querySelector(".bx--tag");
    expect(tag).toBeTruthy();
    expect(getAllByText("GI Upset").length).toEqual(2);

    //remove Tag
    fireEvent.click(container.querySelector(".bx--tag__close-icon"));
    const checkbox = screen.getByLabelText("GI Upset");
    expect(checkbox.checked).toEqual(false);
    expect(getAllByText("GI Upset").length).toEqual(1);
  });

  it("should render SelectReactions with search bar", function () {
    const { container } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    expect(container.querySelector(".bx--search--xl")).not.toBeNull();
  });
  it("should render Common Reactions", function () {
    render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    expect(screen.getByText("Common Reactions")).toBeTruthy();
  });

  it("should render Checkboxes based on Search text", function () {
    const { container } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    const searchInput = container.querySelector(".bx--search-input");
    fireEvent.change(searchInput, { target: { value: "set" } });
    expect(screen.getByText("GI Upset")).toBeTruthy();
    expect(() => screen.getByText("Fever")).toThrowError();
  });

  it("should set isSearchResultEmpty to true and reset searchResults when key is empty", () => {
    const { container } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    const searchInput = container.querySelector(".bx--search-input");

    fireEvent.change(searchInput, { target: { value: "vw" } });

    const allCheckboxes = container.querySelectorAll(".bx--checkbox");
    expect(allCheckboxes).toHaveLength(5); // Assuming mockReactions has 5 reactions
});

  it("should show chiclets for selected reactions", function () {
    const { container, getAllByText } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    selectReaction(container);
    const tag = container.querySelector(".bx--tag");
    expect(tag).toBeTruthy();
    expect(getAllByText("GI Upset").length).toEqual(2);
  });

  it("should set searchKey correctly", () => {
    const { container } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    const searchInput = container.querySelector(".bx--search-input");

    fireEvent.change(searchInput, { target: { value: "GI" } });
    expect(searchInput.value).toBe("GI");
  });

  it("should reset isSearchResultEmpty and searchResults when key is empty", () => {
    const { container, rerender } = render(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );
    const searchInput = container.querySelector(".bx--search-input");

    fireEvent.change(searchInput, { target: { value: "GI" } });
    expect(container.querySelectorAll(".bx--checkbox")).toHaveLength(1);

    fireEvent.change(searchInput, { target: { value: "" } });
    rerender(
      <SelectReactions
        onChange={onChange}
        reactions={mockReactions}
        selectedAllergen={mockSelectedAllergen}
      />
    );

    expect(container.querySelectorAll(".bx--checkbox")).toHaveLength(5);
  });
});
