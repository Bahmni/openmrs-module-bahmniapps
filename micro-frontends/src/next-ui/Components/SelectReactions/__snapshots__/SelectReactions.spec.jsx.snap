// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select reactions should render SelectReactions 1`] = `
<div>
  <div
    class="section-next-ui"
  >
    <div
      class="font-large selected-allergen"
    >
      Selected Allergen:
       
      Eggs
    </div>
    <div
      class="font-large bold"
    >
      Search Reaction
      <span
        class="red-text"
      >
        Â *
      </span>
    </div>
    <div>
      <div
        aria-labelledby="reaction-search-search"
        class="bx--search bx--search--xl"
        role="search"
      >
        <div
          class="bx--search-magnifier"
        >
          <svg
            aria-hidden="true"
            class="bx--search-magnifier-icon"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15,14.3L10.7,10c1.9-2.3,1.6-5.8-0.7-7.7S4.2,0.7,2.3,3S0.7,8.8,3,10.7c2,1.7,5,1.7,7,0l4.3,4.3L15,14.3z M2,6.5	C2,4,4,2,6.5,2S11,4,11,6.5S9,11,6.5,11S2,9,2,6.5z"
            />
          </svg>
        </div>
        <label
          class="bx--label"
          for="reaction-search"
          id="reaction-search-search"
        />
        <input
          autocomplete="off"
          class="bx--search-input"
          id="reaction-search"
          placeholder="Type to search Reactions"
          role="searchbox"
          type="text"
          value=""
        />
        <button
          aria-label="Clear search input"
          class="bx--search-close bx--search-close--hidden"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            focusable="false"
            height="16"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 32 32"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div>
      <div
        class="font-small selected-allergen"
      >
        Common Reactions
      </div>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          class="bx--checkbox"
          id="101AA"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="101AA"
        >
          <span
            class="bx--checkbox-label-text"
            dir="auto"
          >
            GI Upset
          </span>
        </label>
      </div>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          class="bx--checkbox"
          id="102AA"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="102AA"
        >
          <span
            class="bx--checkbox-label-text"
            dir="auto"
          >
            Fever
          </span>
        </label>
      </div>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          class="bx--checkbox"
          id="103AA"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="103AA"
        >
          <span
            class="bx--checkbox-label-text"
            dir="auto"
          >
            Headache
          </span>
        </label>
      </div>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          class="bx--checkbox"
          id="104AA"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="104AA"
        >
          <span
            class="bx--checkbox-label-text"
            dir="auto"
          >
            Nausea
          </span>
        </label>
      </div>
      <div
        class="bx--form-item bx--checkbox-wrapper"
      >
        <input
          class="bx--checkbox"
          id="105AA"
          type="checkbox"
        />
        <label
          class="bx--checkbox-label"
          for="105AA"
        >
          <span
            class="bx--checkbox-label-text"
            dir="auto"
          >
            Cough
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
